Adding plotting features to toolbox

kevindougherty-noaa opened the pull request:
Here I will add plotting scripts that can be used to create diagnostics. 

One thing I will need is to add a the NOAA/NWS label to the plots. Also, @CoryMartin-NOAA I think this ([pyshp](https://github.com/GeospatialPython/pyshp)) is where the shapefiles I need to download to utilize the cartopy features, can you confirm for me? 

https://scitools.org.uk/cartopy/docs/latest/tutorials/using_the_shapereader.html

kevindougherty-noaa committed:
initial commit

kevindougherty-noaa committed:
initial commit; sptial plots

kevindougherty-noaa committed:
Delete units.py

kevindougherty-noaa committed:
Delete __init__.py

kevindougherty-noaa committed:
Delete wind.py

kevindougherty-noaa committed:
Merge branch 'develop' into feature/plotting

CoryMartin-NOAA commented:
@kevindougherty-noaa can you try to set `cartopy.config['data_dir']` to something and see if it tries to find shapefiles there?

aerorahul commented on src/emc/plots/spatial.py:
```diff
@@ -0,0 +1,136 @@
+from cartopy.mpl.ticker import LongitudeFormatter, LatitudeFormatter
+
+
+def _global(latitude, longitude, data, **kwargs):
```
if you are going to use `**kwargs`, you should check if that arg is present when you use it.
If not present, it should provide a default or error out.
Also, instead of `**kwargs`, I suggest with pass a dictionary.  The dictionary is empty by default.

aerorahul replied:
```suggestion
def _global(latitude, longitude, data, plotopts=dict()):
```

CoryMartin-NOAA replied:
@kevindougherty-noaa and I talked in length about the dictionary vs args issue in these functions.

I agree it probably needs to check if it exists. However, he convinced me that it is more work/less clean to have a dictionary of options passed into these functions than just making them optional arguments.

aerorahul replied:
you could have both, `plotopts=dict()` and `**kwargs`

aerorahul replied:
The rationale for a dictionary comes from this:
https://github.com/aerorahul/pyarsenal/blob/develop/src/pyarsenal/specs.py

We can store a database of `field_specs`, that populates and gives a default dictionary of values.

aerorahul commented on src/emc/plots/spatial.py:
```diff
@@ -0,0 +1,136 @@
+    lat_formatter = LatitudeFormatter()
+    ax.xaxis.set_major_formatter(lon_formatter)
+    ax.yaxis.set_major_formatter(lat_formatter)
+
```
```suggestion
vmax = plotopts.get('vmax', data.max())
vmax = plotopts.get('vmax', data.min())

```

CoryMartin-NOAA replied:
@kevindougherty-noaa I forgot about this syntax that @aerorahul pointed out, this is probably the way to go, it checks if the entry is in the dict, and if not, sets it to this value.

aerorahul commented on src/emc/plots/spatial.py:
```diff
@@ -0,0 +1,136 @@
+    ax.yaxis.set_major_formatter(lat_formatter)
+
+    cs = plt.scatter(longitude, latitude, c=data, s=10,
+                     vmin=kwargs['vmin'], vmax=kwargs['vmax'],
```
```suggestion
                     vmin=vmin, vmax=vmax,
```

aerorahul commented on src/emc/plots/spatial.py:
```diff
@@ -0,0 +1,136 @@
+
+    cs = plt.scatter(longitude, latitude, c=data, s=10,
+                     vmin=kwargs['vmin'], vmax=kwargs['vmax'],
+                     cmap=kwargs['cmap'], transform=ccrs.PlateCarree())
```
```suggestion
                     cmap=plotopts.get('cmap','jet'), transform=ccrs.PlateCarree())
```

aerorahul reviewed (changes requested):
I am making some suggestions.
please take a look

kevindougherty-noaa committed:
updates to address comments; split map projections into functions; created functions that plot 1D/2D data

kevindougherty-noaa committed:
switch lon and lat in _spatial_scatter

kevindougherty-noaa commented:
@aerorahul I addressed your comments and added a `plotopts` dictionary so that users cannot just add whatever they want as **kwargs. Cory and I discussed yesterday as well for this script to be able to handle both 1 dimensional and 2 dimensional data. Doing so, I created domain functions so that in the future, if an additional domain needs to be included, we would just have to add a new function and an additional conditional.

Here are some examples using 1D wind data:

`from netCDF4 import Dataset

fname = "/scratch2/NCEPDEV/stmp1/Emily.Liu/data_aop20_global/SATWND/2020110112/obs/satwind_avhrr_obs_v2_2020110112.nc4"

with Dataset(fname, mode='r') as f:
   
    latitude = f.groups['MetaData'].variables['latitude'][:]
    longitude = f.groups['MetaData'].variables['longitude'][:]
    u = f.groups['ObsValue'].variables['eastward_wind'][:]
    v = f.groups['ObsValue'].variables['eastward_wind'][:]
    
# calculate data statistics
windspeed = np.sqrt(np.square(u) + np.square(v))

fig = spatial(latitude, longitude, windspeed, domain='global', cmap='viridis',
              title='test 1D data - wind speed')`

![image](https://user-images.githubusercontent.com/69815622/126809518-60bad1c1-7dcd-497e-acfe-eb8cdf739f2b.png)

2D temperature data:

`from netCDF4 import Dataset

file = '/scratch1/NCEPDEV/da/Cory.R.Martin/ROTDIRS/v16p1ctl/gdas.20210614/00/atmos/gdas.t00z.atmf006.nc'

with Dataset(file, mode='r') as f:
    lat = f.variables['lat'][:]
    lon = f.variables['lon'][:]
    temp = f.variables['tmp'][:]
    
fig = spatial(lat, lon, temp[0,126,:,:], domain='global', cmap='jet',
              vmax=310, vmin=200, grid=True, title='test 2D data - temperature')`

![image](https://user-images.githubusercontent.com/69815622/126809664-10aebffd-30a8-4961-960a-340ad4f246b8.png)

kevindougherty-noaa committed:
added addtional parameters to description

aerorahul commented on src/emc/plots/spatial.py:
```diff
@@ -0,0 +1,160 @@
+    vmax : maximum value of data
+    title : plot title (str)
+    cycle : cycle of data (str)
+    cbar_label : colorbar label
```
we need to work on the syntax of this now before it is too far.
Investigate `pdoc` syntax.

aerorahul commented on src/emc/plots/spatial.py:
```diff
@@ -0,0 +1,160 @@
+    domain = domain.lower()
+
+    # Dictionary of kwargs
+    plotopts = {'cmap': cmap, 'marker size': markersize, 'vmin': vmin,
```
If `vmin` and/or `vmax` are `None`, they  should be calculated from `data`.
Use `utils/roundNumber` to get a nice round number for the min and max.
If `title` is `None`, create some arbitrary string e.g. `EMCpy Spatial Plot` or whatever.

kevindougherty-noaa replied:
Okay. Vmin and vmax when set to None automatically grad the max and min of the data. You would prefer a rounded number? Also, that's a great idea for a title, thank you.

aerorahul replied:
I think so.  If the vmin and vmax are the min, max of data, and those values are e.g. 13.12315, 272.1231, the colorbar looks whacky IMO.
I would rather compute the min, max from the data and then round it to the nearest appropriate number.
It just looks better.

aerorahul commented on src/emc/plots/spatial.py:
```diff
@@ -0,0 +1,160 @@
+    plotopts = {'cmap': cmap, 'marker size': markersize, 'vmin': vmin,
+                'vmax': vmax, 'grid': grid, 'title': title, 'cycle': cycle,
+                'cbar label': cbar_label}
+
+    # Get domain fig/ax
+    if domain == 'global':
+        fig, ax = _global()
+
+    elif domain == 'conus':
+        fig, ax = _conus()
+
+    elif domain == 'north america':
+        fig, ax = _north_america()
+
+    else:
+        raise TypeError(f'{domain} is not a valid domain. Current ' +
+                        'domains supported are: global, conus, north america')
```
I think this is ripe for a factory.
I can try and come up with an implementation in a later PR.

aerorahul commented on src/emc/plots/spatial.py:
```diff
@@ -0,0 +1,160 @@
+from matplotlib import rcParams, ticker, cm
+import cartopy.feature as cfeature
+import cartopy.crs as ccrs
+import matplotlib.colors as mcolors
```
please group together with other matplotlib imports.
Might help with some consistency too e.g.
`from matplotlib import colors as mcolors`

kevindougherty-noaa replied:
I realized I wasn't even using most of them so I will just delete in next commit.

kevindougherty-noaa committed:
updated doc string

kevindougherty-noaa committed:
added roundNumber to vmin/vmax; moved to emcpy/plots

kevindougherty-noaa committed:
files were not added in last commit

kevindougherty-noaa committed:
Delete spatial_notebook.ipynb

kevindougherty-noaa committed:
Delete scatter_notebook.ipynb

kevindougherty-noaa committed:
fixed to pass pycode style

kevindougherty-noaa committed:
Merge branch 'feature/plotting' of https://github.com/NOAA-EMC/emcpy into feature/plotting

kevindougherty-noaa commented:
Closing this PR to update and reorganize the code to make cleaner in a new PR.
